### ISSUE TITLE
NPY support

### ISSUE THREAD

**Issue by rossant** [blaze/dask#684]
It would be nice to have a method `array.to_npy('file.npy')`

**Comment by mrocklin** [comment]
Sure.  Just to be clear, my understanding is that npy files do not support partial reads so we should be able to assume that the result fits in memory.  Is this correct?

If this is the case then the implementation will probably be something only marginally more complex than the following:

```python
def save(fn, x):
    np.save(fn, x.compute()
```

**Comment by rossant** [comment]
Actually I would need this for huge out-of-memory arrays. AFAIU [npy does support memmapping](http://docs.scipy.org/doc/numpy/reference/generated/numpy.load.html)?

To write a big array on disk maybe you could first write an empty array in a file, and fill it chunk by chunk? Otherwise there might be a way to just append data to an existing file, since [npy](http://docs.scipy.org/doc/numpy-dev/neps/npy-format.html) is just a flat binary file with a header.

**Comment by mrocklin** [comment]
Ah, I had forgotten about memmap.

Your suggested approach should work well.  Here is a small test:

```python
def test_memmap():
    with tmpfile('npy') as fn_1:
        with tmpfile('npy') as fn_2:
            x = da.arange(100, chunks=15)
            target = np.memmap(fn_1, shape=x.shape, mode='w+', dtype=x.dtype)

            x.store(target)
            assert eq(target, x)

            np.save(fn_2, target)
            assert eq(np.load(fn_2, mmap_mode='r'), x)
```

**Comment by mrocklin** [comment]
Hrm, this approach unfortunately involves two complete writes to disk.

**Comment by jcrist** [comment]
From the format spec, we could probably do an append approach if we wanted (https://github.com/numpy/numpy/blob/master/doc/neps/npy-format.rst). Would you want the arrays written as chunks (format looks like multiple arrays on disk to numpy), or as a single array (looks like 1 array on disk to numpy)?

**Comment by rossant** [comment]
@jcrist probably multiple arrays on disk. Here's what I'm trying to do:

Initially I have a dask array `D=(D1, D2, ..., Dn)` (2D arrays chunked along axis=1) loaded from a big flat binary array. Then a function `f` is applied to every chunk in parallel, returning array `O=(O1, ..., On)`. I want to persist this big output array on disk, using the `npy` format, in a sensible way that I can control. For example, something like `.dask_cache/output_array_1.npy` etc. AFAIK we can't write to a single big file because of file concurrency issues with multiprocessing.

There would actually be several tasks like this, caching results in different subfolders.

These files would be used in two ways:

* if you relaunch the task, or if the process crashed in the middle, the existing results are directly reused
* the user may want to export the output to a big `npy` array, concatenation of the chunked `npy` files

**Comment by mrocklin** [comment]
Ah, ok, so you don't want a single npy file, you're looking for a more custom/chunked solution.  Implementing this is probably easier than figuring out the right API/spec.  What were you thinking?  Perhaps the following?

```python
x.to_npy(dirname, axis=0)
```

**Comment by rossant** [comment]
Yes, something like that. I guess there would be one directory per dask array, with one file per chunk?

Conversely, how would I load this set of files into a dask array without loading all files into memory? This would be to export this multi-npy-files array into a single npy array.

**Comment by mrocklin** [comment]
I was thinking one file per slice rather than one file per chunk.  

Regarding loading, this work would accompany a second function to load.

**Comment by mrocklin** [comment]
So @rossant just to be clear on things before I start on this.  Does the following look right to you?

```python
>>> da.to_npy_stack('/path/to/data/', my_dask_array, axis=0)
>>> y = da.from_npy_stack('/path/to/data/')

$ ls /path/to/data/
0.npy
1.npy
2.npy
```

**Comment by shoyer** [comment]
Just a note -- reading mmapped arrays with dask is already highly performant with the multi-threaded engine. We do this all the time for reading netcdf3 files in xray via scipy (which uses mmap for reads).

**Comment by rossant** [comment]
@mrocklin to be clear can you precise the structure of `my_dask_array` and the contents of `0.npy` etc?

**Comment by mrocklin** [comment]
```python
>>> my_dask_array.shape
(5, 1000, 2000)

>>> np.load('/path/to/data/0.npy').shape
(1000, 2000)
```

**Comment by mrocklin** [comment]
```python
>>> (my_dask_array[0].compute() == np.load('/path/to/data/0.npy')).all()
True
```

**Comment by rossant** [comment]
ah, okay, so that's not exactly what I had in mind. In my case I have `my_dask_array.shape == (5000, 2000)` with, say, 100 chunks of size 50, and I'd expect `np.load('/path/to/data/0.npy').shape == (50, 2000)`

**Comment by mrocklin** [comment]
Potential candidate implemented in #686.  Feedback welcome.

As a general disclaimer though this seems like the kind of problem for which several valid solutions exist.  It's entirely possible that this API changes significantly in the future.
